zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1088 @@
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Leaf(right)),
+            },
+            Joined::Three(left, mid, right) => SpanningTree::Parent {
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Parent {
+                    span: mid.block_range().start..right.block_range().end,
+                    left: Box::new(SpanningTree::Leaf(mid)),
+                    right: Box::new(SpanningTree::Leaf(right)),
+                }),
+            },
+        }
+    }
+
+    fn from_insert(
+        left: Box<Self>,
+        right: Box<Self>,
+        to_insert: ScanRange,
+        insert: Insert,
+    ) -> Self {
+        let (left, right) = match insert {
+            Insert::Left => (Box::new(left.insert(to_insert)), right),
+            Insert::Right => (left, Box::new(right.insert(to_insert))),
+        };
+        SpanningTree::Parent {
+            span: left.span().start..right.span().end,
+            left,
+            right,
+        }
+    }
+
+    fn from_split(left: Self, right: Self, to_insert: ScanRange, split_point: BlockHeight) -> Self {
+        let (l_insert, r_insert) = to_insert
+            .split_at(split_point)
+            .expect("Split point is within the range of to_insert");
+        let left = Box::new(left.insert(l_insert));
+        let right = Box::new(right.insert(r_insert));
+        SpanningTree::Parent {
+            span: left.span().start..right.span().end,
+            left,
+            right,
+        }
+    }
+
+    fn insert(self, to_insert: ScanRange) -> Self {
+        match self {
+            SpanningTree::Leaf(cur) => Self::from_joined(insert(cur, to_insert)),
+            SpanningTree::Parent { span, left, right } => {
+                // This algorithm always preserves the existing partition point, and does not do
+                // any rebalancing or unification of ranges within the tree. This should be okay
+                // because `into_vec` performs such unification, and the tree being unbalanced
+                // should be fine given the relatively small number of ranges we should ordinarily
+                // be concerned with.
+                use RangeOrdering::*;
+                match RangeOrdering::cmp(&span, to_insert.block_range()) {
+                    LeftFirstDisjoint => {
+                        // extend the right-hand branch
+                        Self::from_insert(left, right, to_insert, Insert::Right)
+                    }
+                    LeftFirstOverlap => {
+                        let split_point = left.span().end;
+                        if split_point > to_insert.block_range().start {
+                            Self::from_split(*left, *right, to_insert, split_point)
```

Comment:
I think we should pay attention to the check warnings saying that some of these cases are not covered by tests.